fe/fe-core/src/main/java/com/starrocks/statistic/columns/PredicateColumnsStorage.java

```diff
@@ -0,0 +1,319 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.statistic.columns;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import com.starrocks.analysis.TableName;
+import com.starrocks.catalog.ColumnId;
+import com.starrocks.catalog.Database;
+import com.starrocks.catalog.Table;
+import com.starrocks.common.Pair;
+import com.starrocks.common.util.DateUtils;
+import com.starrocks.common.util.FrontendDaemon;
+import com.starrocks.common.util.TimeUtils;
+import com.starrocks.load.pipe.filelist.RepoExecutor;
+import com.starrocks.scheduler.history.TableKeeper;
+import com.starrocks.server.GlobalStateMgr;
+import com.starrocks.server.LocalMetastore;
+import com.starrocks.statistic.StatsConstants;
+import com.starrocks.thrift.TResultBatch;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import org.apache.commons.collections4.ListUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.velocity.VelocityContext;
+import org.apache.velocity.app.VelocityEngine;
+
+import java.io.StringWriter;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * 1. Persistence:
+ * The in-memory state is periodically persisted, so we use the ColumnUsage.lastUsage and this.lastPersist to
+ * identify whether one ColumnUsage needs to be persisted.
+ * 2. Query:
+ * The in-memory state can only represent local state, the persisted state is shared by all fe nodes
+ * 3. Vacuum:
+ * If the ColumnUsage.lastUsage is less than now() - ttl, it would be removed from memory state and persisted state
+ * 4. Restore:
+ * Before restoring persisted state into memory, some queries may already come up, so we need to merge these two
+ * kinds of state. Before restoring state we cannot persist otherwise they would conflict
+ */
+public class PredicateColumnsStorage extends FrontendDaemon {
+
+    private static final Logger LOG = LogManager.getLogger(PredicateColumnsStorage.class);
+
+    public static final String DATABASE_NAME = StatsConstants.STATISTICS_DB_NAME;
+    public static final String TABLE_NAME = "predicate_columns";
+    public static final String TABLE_FULL_NAME = DATABASE_NAME + "." + TABLE_NAME;
+    private static final String TABLE_DDL = "CREATE TABLE " + TABLE_NAME + "( " +
+
+            // PRIMARY KEYS
+            "fe_id STRING NOT NULL," +
+            "db_id BIGINT NOT NULL," +
+            "table_id BIGINT NOT NULL," +
```

Comment:
does this support `external table`? external table's table_id is not stable?

Reply:
not supported actually. I cannot figure out a solution to get a int64 id for external table.